cpp/src/arrow/io/compressed_benchmark.cc

```diff
@@ -0,0 +1,166 @@
+
+  bool supports_zero_copy() const override { return false; }
+
+  Result<int64_t> Read(int64_t nbytes, void* out) override {
+    return reader_.Read(nbytes, out);
+  }
+
+  Result<std::shared_ptr<Buffer>> Read(int64_t nbytes) override {
+    // Testing the non-zero copy path like reading from local file or Object store,
+    // so we need to allocate a buffer and copy the data.
+    ARROW_ASSIGN_OR_RAISE(auto buf, ::arrow::AllocateResizableBuffer(nbytes));
+    ARROW_ASSIGN_OR_RAISE(int64_t size, Read(nbytes, buf->mutable_data()));
+    ARROW_RETURN_NOT_OK(buf->Resize(size));
+    return buf;
+  }
+  Status Close() override { return reader_.Close(); }
+  Result<int64_t> Tell() const override { return reader_.Tell(); }
+  bool closed() const override { return reader_.closed(); }
+
+ private:
+  ::arrow::io::BufferReader reader_;
+};
+
+template <typename BufReader, Compression::type COMPRESSION>
+static void CompressionInputBenchmark(::benchmark::State& state) {
+  std::vector<uint8_t> data = MakeCompressibleData(/*data_size=*/state.range(0));
+  int64_t per_read_bytes = state.range(1);
+  auto codec = ::arrow::util::Codec::Create(COMPRESSION).ValueOrDie();
+  int64_t max_compress_len = codec->MaxCompressedLen(data.size(), data.data());
+  std::shared_ptr<::arrow::ResizableBuffer> buf =
+      ::arrow::AllocateResizableBuffer(max_compress_len).ValueOrDie();
+  int64_t length =
+      codec->Compress(data.size(), data.data(), max_compress_len, buf->mutable_data())
+          .ValueOrDie();
+  ABORT_NOT_OK(buf->Resize(length));
+  for (auto _ : state) {
+    state.PauseTiming();
+    auto reader = std::make_shared<BufReader>(buf);
+    auto input_stream =
+        ::arrow::io::CompressedInputStream::Make(codec.get(), reader).ValueOrDie();
+    auto read_buffer = ::arrow::AllocateBuffer(per_read_bytes).ValueOrDie();
+    state.ResumeTiming();
+    int64_t remaining_size = data.size();
+    while (remaining_size > 0) {
+      auto value = input_stream->Read(per_read_bytes, read_buffer->mutable_data());
+      ABORT_NOT_OK(value);
+      remaining_size -= value.ValueOrDie();
+    }
+  }
+  state.SetBytesProcessed(length * state.iterations());
+}
+
+template <Compression::type COMPRESSION>
+static void CompressionInputZeroCopyBenchmark(::benchmark::State& state) {
+  CompressionInputBenchmark<::arrow::io::BufferReader, COMPRESSION>(state);
+}
+
+template <Compression::type COMPRESSION>
+static void CompressionInputNonZeroCopyBenchmark(::benchmark::State& state) {
+  CompressionInputBenchmark<NonZeroCopyBufferReader, COMPRESSION>(state);
+}
+
+static void CompressedInputArguments(::benchmark::internal::Benchmark* b) {
+  b->ArgNames({"InputBytes", "PerReadBytes"})
+      ->Args({8 * 1024, 8 * 1024})
+      ->Args({64 * 1024, 8 * 1024})
+      ->Args({64 * 1024, 64 * 1024})
+      ->Args({1024 * 1024, 8 * 1024})
+      ->Args({1024 * 1024, 64 * 1024})
+      ->Args({1024 * 1024, 1024 * 1024});
+}
+
+#ifdef ARROW_WITH_ZLIB
```

Comment:
I don't think it's worth testing all codecs. We should only test the codec with the smallest overhead, i.e. LZ4.